src/Features/Core/Portable/Completion/Providers/AbstractAwaitCompletionProvider.cs

```diff
@@ -0,0 +1,89 @@
+#pragma warning restore VSTHRD200 // Use "Async" suffix for async methods
+
+        public sealed override async Task ProvideCompletionsAsync(CompletionContext context)
+        {
+            var document = context.Document;
+            var position = context.Position;
+            var cancellationToken = context.CancellationToken;
+            var syntaxFacts = document.GetRequiredLanguageService<ISyntaxFactsService>();
+            var syntaxTree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
+            if (syntaxFacts.IsInNonUserCode(syntaxTree, position, cancellationToken))
+            {
+                return;
+            }
+
+            var semanticModel = await document.ReuseExistingSpeculativeModelAsync(position, cancellationToken).ConfigureAwait(false);
+            var workspace = document.Project.Solution.Workspace;
+            var syntaxContext = document.GetRequiredLanguageService<ISyntaxContextService>().CreateContext(workspace, semanticModel, position, cancellationToken);
+            if (!syntaxContext.IsAwaitKeywordContext())
+            {
+                return;
+            }
+
+            var completionItem = GetCompletionItem(syntaxContext.TargetToken);
+            context.AddItem(completionItem);
+        }
+
+        public sealed override async Task<CompletionChange> GetChangeAsync(Document document, CompletionItem item, char? commitKey = null, CancellationToken cancellationToken = default)
+        {
+            // IsComplexTextEdit is true when we want to add async to the container.
+            if (!item.IsComplexTextEdit)
+            {
+                return await base.GetChangeAsync(document, item, commitKey, cancellationToken).ConfigureAwait(false);
+            }
+
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var declaration = GetAsyncSupportingDeclaration(root.FindToken(item.Span.Start));
+            if (declaration is null)
+            {
+                // IsComplexTextEdit should only be true when GetAsyncSupportingDeclaration returns non-null.
+                // This is ensured by the ShouldMakeContainerAsync overrides.
+                Debug.Assert(false, "Expected non-null value for declaration.");
+                return await base.GetChangeAsync(document, item, commitKey, cancellationToken).ConfigureAwait(false);
+            }
+
+            using var _ = ArrayBuilder<TextChange>.GetInstance(out var builder);
+            builder.Add(new TextChange(new TextSpan(GetSpanStart(declaration), 0), AsyncKeywordTextWithSpace));
+            builder.Add(new TextChange(item.Span, item.DisplayText));
+
+            var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+            var newText = text.WithChanges(builder);
+            return CompletionChange.Create(Utilities.Collapse(newText, builder.ToImmutableArray()));
```

Comment:
i presume this logic moved up effectively unchanged?

Reply:
Yes.